File: python/atelier_core/models/customer.py
"""Customer: gets `reference()` from the HasReference MIXIN CLASS (MRO), prefix overridden to CU."""

from __future__ import annotations

from ..concerns.has_reference import HasReference


class Customer(HasReference):
    reference_prefix = "CU"

    def __init__(self, id: int, name: str, email: str, phone: str | None = None) -> None:
        self.id = id
        self.name = name
        self.email = email
        self.phone = phone

    def display_name(self) -> str:
        return f"{self.name} <{self.email}>"

    def is_reachable(self) -> bool:
        return self.phone is not None or self.email != ""

    @classmethod
    def seed(cls, id: int, name: str, email: str, phone: str | None = None) -> "Customer":
        """Alternate constructor: the reference number follows the id."""
        customer = cls(id, name, email, phone)
        customer.reference_number = id
        return customer

File: go/core/models/customer.go
// Package models is the domain: the structs the whole corpus asks questions
// about. Mutable models use POINTER receivers; value objects use value ones.
package models

import (
	"fmt"
	"strings"

	"atelier.example/lane/core/concerns"
	"atelier.example/lane/core/support"
)

// Customer EMBEDS concerns.HasReference: Reference() and ShortReference() are
// promoted onto Customer even though neither name appears in this file again.
type Customer struct {
	concerns.HasReference
	ID    int
	Name  string
	Email string
	Phone string
}

// SeedCustomer is the frozen-seed constructor: it fills the embedded reference
// state, which is what makes customer.Reference() answer "CU-2026-000001".
func SeedCustomer(id int, name string, email string, phone string) *Customer {
	return &Customer{
		HasReference: concerns.NewHasReference(support.CustomerPrefix, id),
		ID:           id,
		Name:         name,
		Email:        email,
		Phone:        phone,
	}
}

// IsReachable is true when a phone number was recorded.
func (c *Customer) IsReachable() bool {
	return strings.TrimSpace(c.Phone) != ""
}

// DisplayName is the name plus the masked mail domain.
func (c *Customer) DisplayName() string {
	_, domain, found := strings.Cut(c.Email, "@")
	if !found {
		return c.Name
	}
	return fmt.Sprintf("%s <%s>", c.Name, domain)
}
